---
title: "&nbsp;v1.5"
output:
  prettydoc::html_pretty:
    theme: architect
    highlight: github
    keep_md: no
    includes:
        before_body: histomx_header.html
        after_body: histomx_footer.html
params:
  rcc_file: 'missing-filename'
  rna_file: ''
  patient_file: ''
---

```{r, include=FALSE, eval=FALSE}
## to run manually
rmarkdown::render('./scripts/histomx_report.Rmd')
## convert output html to paged PDF 
system('wkhtmltopdf --javascript-delay 1 ~/Desktop/histomx_report.html ~/Desktop/histomx_report.pdf')
```

```{r setup, include=FALSE, echo=FALSE}
options(rgl.useNULL=TRUE)
## set paths in all code chunks are relative to this root.dir
rdir=getwd()
knitr::opts_knit$set(rdir)

library(htmltools, quietly=TRUE)
library(ggplot2, quietly=TRUE)
theme_set(theme_minimal())
library(gridExtra, quietly=TRUE)
library(knitr, quietly=TRUE)
opts_knit$set(eval.after = "fig.cap")
library(kableExtra, quietly=TRUE)
#library(xtable)
knitr::opts_chunk$set(echo = TRUE)
```

***
## Molecular Pathology Platform
> **Team leader:** Pr Alexandre Loupy MD/PhD  
**Project managers:** Fariza Mezine & Blaise Robin  
**Bioinformatician:** Dina Zielinski PhD
**Clinical reviewer:** Valentin Goutaudier MD  
**Contact:** jessy.dagobert@inserm.fr / +33-1-53-98-80-85  
> Inserm U970 - 56 rue Leblanc 75015 Paris, France

```{r, echo=FALSE, warning=FALSE, include=FALSE, eval=TRUE}
## import functions
source('../scripts/BHOT.R')
## import reference set, normalize new sample with reference samples and output predictions
source('../scripts/BHOTpred.R')
```

```{r, echo=FALSE, warning=FALSE, include=FALSE, eval=TRUE}
## import parameters (list of input files)
## see render-rmarkdown.sh to run markdown with new RCC as input
newPatient <- params$patient_file
newRNA <- params$rna_file
newRCC <- params$rcc_file

newPath <- normalizePath(dirname(newRCC))
dir.create(paste0(newPath, '/histomx_results/'), recursive=TRUE, showWarnings=FALSE)
outPath=paste0(newPath, '/histomx_results/')

```

```{r, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
## the following functions output all files to a directory called 'histomx_results' in the input file path
## and return a list of these files
newPreds <- BHOTpred(newRCC, outPath, saveFiles=FALSE)
newID <- gsub("-", ".", newPreds$score_table$ID)
dev.off()

## Output sequencing run info
newQC <- rccQC(newRCC, outPath)

```

***
## Patient info:
```{r, echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}

if(!is.na(file.info(newPatient)$size)) {
  
    ## replace empty cells with "NA"
    pt_df <- read.table(newPatient, sep='\t', na.strings=c("", "NA"))
    ## exclude empty cells
    pt_df <- pt_df %>% filter_all(all_vars(!is.na(.)))
    #newID=pt_df$'ID center'
      
} else {
    ## display empty data frame
    pt_df <- data.frame('Platform ID'=newPreds$score_table$ID,
                        'Date of transplantation'=" ", 'Date of biopsy'=" ",
                        'Biopsy indication'="    ", check.names=FALSE)
    pt_df <- data.frame(V1=rownames(t(pt_df)), V2=t(pt_df)[,1])
  }

## format and output patient info table
if(nrow(pt_df) %% 2 == 0) {
    pt_df <- cbind(pt_df[1:(nrow(pt_df)/2),], pt_df[(nrow(pt_df)/2+1):nrow(pt_df),])
    kable(pt_df, row.names=FALSE, col.names=rep("", ncol(pt_df)), align='c', format="html") %>%
          row_spec(1:nrow(pt_df), background="white") %>%
          column_spec(c(1,3), bold=TRUE, background="whitesmoke")
} else {
    ## todo: set 1st row as single row, remaining as double
    kable(pt_df, row.names=FALSE, col.names=rep("", ncol(pt_df)), align='c', format="html") %>%
    row_spec(1:nrow(pt_df), background="white") %>%
    column_spec(1, bold=TRUE, background="whitesmoke")
}

```

***
## Gene expression based probabilities
***
### Diagnosis based probabilities (%):
```{r, echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}
## molecular score table
ms_new <- data.frame(t(newPreds$score_table[,c("AMR", "TCMR", "ATI", "IFTA")]))
## exclude 'normal' molecular score

## reference set quantiles
ms_ref <- newPreds$ref_scores[,c("Dx", "AMR", "TCMR", "ATI", "normal", "IFTA", "Q")]
 
## q1 and q3 (can also add IQR, min, max)
#tab <- data.frame(matrix(nrow=dim(ms_new)[1], ncol=length(unique(ms_ref$Dx))*2))
#colnames(tab) <- c("AMR Q1", "AMR Q3", "TCMR Q1", "TCMR Q3", "ATI Q1", "ATI Q3", "No lesion Q1", "No lesion Q3")

tab <- data.frame(matrix(nrow=dim(ms_new)[1], ncol=2))
colnames(tab) <- c("No lesion Q1", "No lesion Q3")
rownames(tab) <- rownames(ms_new)

## rows=molecular scores; columns=histo diagnosis
## reference biopsy AMR molecular score ranges
# tab[rownames(tab)=="AMR","AMR Q1"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="25%","AMR"]
# tab[rownames(tab)=="AMR","AMR Q3"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="75%","AMR"]
# tab[rownames(tab)=="AMR","TCMR Q1"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="25%","AMR"]
# tab[rownames(tab)=="AMR","TCMR Q3"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="75%","AMR"]
# tab[rownames(tab)=="AMR","ATI Q1"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="25%","AMR"]
# tab[rownames(tab)=="AMR","ATI Q3"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="75%","AMR"]
tab[rownames(tab)=="AMR","No lesion Q1"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="25%","AMR"]
tab[rownames(tab)=="AMR","No lesion Q3"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="75%","AMR"]

## reference biopsy acute TCMR molecular score ranges
# tab[rownames(tab)=="TCMR","AMR Q1"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="25%","TCMR"]
# tab[rownames(tab)=="TCMR","AMR Q3"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="75%","TCMR"]
# tab[rownames(tab)=="TCMR","TCMR Q1"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="25%","TCMR"]
# tab[rownames(tab)=="TCMR","TCMR Q3"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="75%","TCMR"]
# tab[rownames(tab)=="TCMR","ATI Q1"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="25%","TCMR"]
# tab[rownames(tab)=="TCMR","ATI Q3"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="75%","TCMR"]
tab[rownames(tab)=="TCMR","No lesion Q1"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="25%","TCMR"]
tab[rownames(tab)=="TCMR","No lesion Q3"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="75%","TCMR"]

## reference biopsy ATI molecular score ranges
tab[rownames(tab)=="ATI","No lesion Q1"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="25%","ATI"]
tab[rownames(tab)=="ATI","No lesion Q3"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="75%","ATI"]

## reference biopsy IFTA molecular score ranges
tab[rownames(tab)=="IFTA","No lesion Q1"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="25%","IFTA"]
tab[rownames(tab)=="IFTA","No lesion Q3"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="75%","IFTA"]

tab <- tab[!is.na(colnames(tab))]
#tab$'AMR range' <- paste(tab$`AMR Q1`, "-", tab$`AMR Q3`)
#tab$'TCMR range' <- paste(tab$`TCMR Q1`, "-", tab$`TCMR Q3`)
#tab$'TCMR range' <- paste(tab$`ATI Q1`, "-", tab$`ATN Q3`)
tab$'Normal range' <- paste(tab$`No lesion Q1`, "-", tab$`No lesion Q3`)
tab$'Diagnosis' <- rownames(tab)

#dx_tab <- tab[tab$'Molecular score' != "No specific Dx",] ## exclude 'normal' molecular score
dx_tab <- cbind(ms_new, tab[,c("Diagnosis", "Normal range")])
dx_tab <- dx_tab %>% relocate('Diagnosis', .before = colnames(dx_tab)[1])

dx_tab$Interpretation <- ifelse(dx_tab[,newID]<=10,"unlikely",
                                       ifelse(dx_tab[,newID]>10 & dx_tab[,newID]<=20,"low probability",
                                              ifelse(dx_tab[,newID]>20 & dx_tab[,newID]<=40,"moderate probability",
                                                     ifelse(dx_tab[,newID]>40 & dx_tab[,newID]<=60,"moderate/high probability",
                                                           ifelse(dx_tab[,newID]>60 & dx_tab[,newID]<=80,"high probability",
                                                                  ifelse(dx_tab[,newID]>80,"very high probability","NA"))))))

kable(dx_tab, row.names=FALSE, align='c', caption="", format="html") %>%
    column_spec(1, bold=TRUE, color="black") %>%
    column_spec(2, bold=TRUE, color="dodgerblue") %>%
    row_spec(1, background="whitesmoke") %>% row_spec(2, background="white")

```
Diagnosis based scores represent the probability that a biopsy has an expression profile similar to reference samples with a given histology based diagnosis: AMR, TCMR, ATI (Acute tubular injury without rejection), atrophy/fibrosis (isolated IFTA).

\newpage

### Banff lesion based probabilities (%):
```{r, echo=FALSE, warning=FALSE, message=FALSE, include=TRUE}
## table of molecular scores
ms_new <- data.frame(t(newPreds$score_table[,c("g0_score", "ptc0_score", "cg0_score", "i1_score", "t1_score", "iifta0_score", "v0_score", "cv1_score", "ci1_score", "ct1_score")]))

## reference set quantiles
ms_ref <- newPreds$ref_scores[,c("Dx", "g0_score", "ptc0_score", "cg0_score", "i1_score", "t1_score", "iifta0_score", "v0_score", "cv1_score", "ci1_score", "ct1_score", "Q")]
 
## q1 and q3 (can also add IQR, min, max)
tab <- data.frame(matrix(nrow=dim(ms_new)[1], ncol=length(unique(ms_ref$Dx))*2))
colnames(tab) <- c("AMR Q1", "AMR Q3", "TCMR Q1", "TCMR Q3",
                   "ATI Q1", "ATI Q3", "IFTA Q1", "IFTA Q3",
                   "No lesion Q1", "No lesion Q3")
rownames(tab) <- rownames(ms_new)

## add reference biopsy molecular score ranges
## rows=molecular scores; columns=histo diagnosis
for(i in 1:nrow(tab)) {
  tab[rownames(tab)==rownames(ms_new)[i],"AMR Q1"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="25%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"AMR Q3"] <- ms_ref[ms_ref$Dx=="AMR" & ms_ref$Q=="75%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"TCMR Q1"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="25%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"TCMR Q3"] <- ms_ref[ms_ref$Dx=="TCMR" & ms_ref$Q=="75%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"ATI Q1"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="25%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"ATI Q3"] <- ms_ref[ms_ref$Dx=="ATI" & ms_ref$Q=="75%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"IFTA Q1"] <- ms_ref[ms_ref$Dx=="IFTA" & ms_ref$Q=="25%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"IFTA Q3"] <- ms_ref[ms_ref$Dx=="IFTA" & ms_ref$Q=="75%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"No lesion Q1"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="25%",rownames(ms_new)[i]]
  tab[rownames(tab)==rownames(ms_new)[i],"No lesion Q3"] <- ms_ref[ms_ref$Dx=="No specific Dx" & ms_ref$Q=="75%",rownames(ms_new)[i]]
}

#tab <- tab[!is.na(colnames(tab))]
tab$'AMR range' <- paste(tab$`AMR Q1`, "-", tab$`AMR Q3`)
tab$'TCMR range' <- paste(tab$`TCMR Q1`, "-", tab$`TCMR Q3`)
tab$'ATI range' <- paste(tab$`ATI Q1`, "-", tab$`ATI Q3`)
tab$'IFTA range' <- paste(tab$`IFTA Q1`, "-", tab$`IFTA Q3`)
tab$'Normal range' <- paste(tab$`No lesion Q1`, "-", tab$`No lesion Q3`)
tab$'Banff lesion' <- gsub("_", " ", rownames(tab))

if(!all(rownames(ms_new)==rownames(tab))) {
  stop("check concordance of lesion score names between new sample and refset")
}

tab <- cbind(ms_new, tab)

tab$Interpretation <- ifelse(tab[,newID]<=10,"unlikely",
                                       ifelse(tab[,newID]>10 & tab[,newID]<=20,"low probability",
                                              ifelse(tab[,newID]>20 & tab[,newID]<=40,"moderate probability",
                                                     ifelse(tab[,newID]>40 & tab[,newID]<=60,"moderate/high probability",
                                                           ifelse(tab[,newID]>60 & tab[,newID]<=80,"high probability",
                                                                  ifelse(tab[,newID]>80,"very high probability","NA"))))))

## return range for 'normal' biopsies
ref_quantiles <- tab[,c(newID, "Banff lesion", "Normal range", "Interpretation")]

ref_quantiles$'Banff lesion' <- gsub("0 score", " > 0", ref_quantiles$'Banff lesion')
ref_quantiles$'Banff lesion' <- gsub("1 score", " > 1", ref_quantiles$'Banff lesion')

ref_quantiles <- ref_quantiles %>% relocate('Banff lesion', .before = colnames(ref_quantiles)[1])
#ref_quantiles <- ref_quantiles[order(ref_quantiles[,newID], decreasing=TRUE),]

kable(ref_quantiles, row.names=FALSE, align='c', format="html") %>%
    column_spec(1, bold=TRUE, color="black") %>%
    column_spec(2, bold=TRUE, color="dodgerblue") %>%
    row_spec(c(1,3,5), background="whitesmoke") %>% row_spec(c(2,4,6), background="white")

#Ranges are defined by Q1 and Q3, the upper and lower bounds of the middle 50% of the reference biopsy molecular scores for each histology based diagnosis.
```
Banff lesion based molecular scores represent the probability that a biopsy has a Banff score >0 or >1, as indicated for each lesion.  

__Interpretations:__   
unlikely (<10%), low (10-20%), moderate (20-40%),  
moderate/high (40-60%), high (60-80%), very high (80-100%)

\newpage

***
## Principal Components Analysis (PCA) of molecular scores:
<div style="margin-bottom:10px;">
```{r, echo=FALSE, warning=FALSE, comment=FALSE, results='asis', fig.align='center', fig.cap='Samples are labeled by histologic diagnosis. The tested sample is the yellow diamond.'}

#gridExtra::grid.arrange(newPreds$pca_1_2, newPreds$pca_2_3, ncol=2)
newPreds$pca_1_2
#newPreds$pca_2_3

```
</div>

***
### Histologic Diagnosis of 50 Nearest Neighbors
```{r, echo=FALSE, warning=FALSE, results='asis', eval=TRUE}
newPreds$knn$Dx <- rownames(newPreds$knn)
newPreds$knn <- newPreds$knn[,c("Dx", "Total", "Percent")]

kable(newPreds$knn, row.names=FALSE, align='c', caption="", format="html") %>%
    column_spec(c(1,2,3), bold=TRUE, background="white")
```
&nbsp;  

\newpage
***
### Pathway analysis
```{r, echo=FALSE, warning=FALSE, comment=FALSE, results='asis', fig.align='center', fig.cap='Genes with increased expression in the tested biopsy compared to all normal biopsies associated with the listed pathway (numbered dots indicate the # of enriched genes; GeneRatio=the % of total genes associated with the pathway).'}

#gridExtra::grid.arrange(newPreds$cell_type_radar, newPreds$pathway_radar, ncol=2, nrow=1)

newPreds$pathway_radar

# pw_df <- newPreds$pathways
# pw_df <- cbind(pw_df[1:(nrow(pw_df)/2),], pw_df[(nrow(pw_df)/2+1):nrow(pw_df),])
# kable(pw_df, row.names=FALSE, col.names=rep("", ncol(pw_df)), align='c', format="html") %>%
#           row_spec(1:nrow(pw_df), background="white") %>%column_spec(c(1,3), bold=TRUE, background="whitesmoke")

```

***
### Cell type analysis
```{r, echo=FALSE, warning=FALSE, comment=FALSE, results='asis', fig.align='center', fig.cap='Genes with increased expression in the tested biopsy compared to all normal biopsies associated with the listed cell type.'}

newPreds$cell_type_radar

# ct_df <- newPreds$cell_types
# ct_df <- cbind(ct_df[1:(nrow(ct_df)/2),], ct_df[(nrow(ct_df)/2+1):nrow(ct_df),])
# kable(ct_df, row.names=FALSE, col.names=rep("", ncol(ct_df)), align='c', format="html") %>%
#           row_spec(1:nrow(ct_df), background="white") %>%column_spec(c(1,3), bold=TRUE, background="whitesmoke")

```

***
### Clinician signout
```{r, echo=FALSE, warning=FALSE, include=TRUE}
sign_df <- data.frame(Name=">", Signature="")

kable(sign_df, align='c', format="html") %>% kableExtra::kable_styling(position="center") %>%
  row_spec(1, background="white", align='l')
```
> **_Comments:_**  
>  &nbsp;    
>  &nbsp;  
>  &nbsp;  
>  &nbsp;    
>  &nbsp;  

\newpage

***
## Technical details
***

### RNA quality
```{r, echo=FALSE, warning=FALSE, include=TRUE}

if(!is.na(file.info(newRNA)$size)) {
  
  rna_df <- read.table(newRNA, sep='\t')

  ## transpose and use first column as header
  if(nrow(rna_df) > 2) {
    rna_df <- setNames(data.frame(t(rna_df[,-1])), rna_df[,1])
  }

  ## if units (ng/ul) included in RNA concentration strip; else convert to numeric
  if(length(strsplit(rna_df$concentration, " ")[[1]]) > 1) {
    rna_df$concentration <- as.numeric(strsplit(rna_df$concentration, " ")[[1]][1])
  } else {
    rna_df$concentration <- as.numeric(rna_df$concentration)
  }

  ## move concentration to end
  rna_df <- rna_df %>% relocate(concentration, .after=last_col())

  ## apply threshold based text color and add back units
  if(rna_df$concentration >= 20) {
      rna_df$status <- "Passed"
      rna_df$concentration <- paste(rna_df$concentration, "ng/ul")
      status_color="green"
  } else if(rna_df$concentration>=10 & rna_df$concentration<20) {
      rna_df$status <- "Caution"
      rna_df$concentration <- paste(rna_df$concentration, "ng/ul")
      status_color="goldenrod"
  } else if(rna_df$concentration < 10) {
      rna_df$status <- "Warning"
      rna_df$concentration <- paste(rna_df$concentration, "ng/ul")
      status_color="firebrick"
  }

  kable(rna_df, row.names=FALSE, align='c', caption="", format="html") %>%
    column_spec(1, bold=TRUE, color="dodgerblue") %>%
    column_spec(ncol(rna_df), bold=TRUE, color=status_color) %>%
    row_spec(1, background="white")
  
} else {
  ## display empty data frame
  rna_df <- data.frame(ID=newPreds$score_table$ID, '260/280 ratio'="", '260/230 ratio'="",
                       concentration="", status="", check.names=FALSE)
  
  kable(rna_df, row.names=FALSE, align='c', caption="", format="html") %>%
      column_spec(1, bold=TRUE, color="dodgerblue") %>%
      column_spec(ncol(rna_df), bold=TRUE) %>%
      row_spec(1, background="white")
}

```
"Caution" status indicates a concentration within the minimum recommended 10-20ng/ul range  
"Warning" status indicates a concentration lower than the recommended minimum 10ng/ul

### NanoString run information
```{r, echo=FALSE, warning=FALSE, include=TRUE}
newQC$qc_table$variable <- rownames(newQC$qc_table)
#newQC$qc_table <- newQC$qc_table[!rownames(newQC$qc_table) %in% "POS_E counts",]

qcTab <- newQC$qc_table[,c("variable", newID)]
qcTab <- qcTab[!qcTab$variable %in% c("FovCount", "FovCounted", "ncgMean", "ncgSD", "FileVersion"),]

qcTab <- cbind(qcTab[1:(nrow(qcTab)/2),], qcTab[(nrow(qcTab)/2+1):nrow(qcTab),])

kable(qcTab, row.names=FALSE, col.names=rep("", ncol(qcTab)), align='l', caption="", format="html") %>%
  column_spec(c(1,3), bold=TRUE, background="whitesmoke") %>%
  column_spec(c(2,4), background="white")

#The version of the RLF file must match the version of the CodeSet used in the hybridization
#counts of POS_E should be higher than 2x the SD above the mean of the negative control
#https://www.nanostring.com/wp-content/uploads/2020/12/Gene_Expression_Data_Analysis_Guidelines.pdf
```
FoV=fields of view; geo mean=geometric mean; LoD=limit of detection; PCL=positive control linearity

__Recommended thresholds:__  
FoV >75%  
Binding density 0.05 to 2.25  
PCL > 0.95  
80% ENDO genes > LoD  
POS_E counts should be > LoD

### For research use only
Models have not been validated in clinical settings or trials. This report does not replace a standard diagnosis. Due to known technical variation, RNA later-fixed samples should be interpreted with caution.

```{r, echo=FALSE, include=TRUE}
#logoIMG='../static/multi-logos.png'
logoIMG='../static/ptg-logo.png'

htmltools::img(src = knitr::image_uri(logoIMG), 
               alt = 'logo', 
               style="display: block; margin-left: auto; margin-right: auto;")
```

